### PR TITLE
Documentation cleanup and improvements

### PR DESCRIPTION
**Pull request by sjpfenninger**
# Summary of changes in this pull request

Documentation cleanup and improvements

* Restructure overall table of contents
* Split existing advanced content into "Advanced constraints" and "Advanced configuration"
* Clean up group constraint documentation and change national-scale example model to use group constraints rather than the now deprecated group_share constraints
* Address #151, #159, #162, #163, #193 (partial), #203

## Bugs fixed

* Fix a bug in group constraint implementation that led to Calliope crashing if a group constraint was defined over all regions (by not setting locs:) but some of techs considered by the consstraint did not exist across all locations
* Fix a bug in group constraint implementation that meant only supply, not supply_plus technologies were considered

The existence of these bugs demonstrates that group constraints need better testing (#193).

## Resolves the following issues

* #151 -- Document use for random seed
* #159 -- Document ``resource_area`` constraints in the new "Advanced constraints" section
* #162 -- Update and move time series documentation into the "building a model" section
* #163 -- Add new troubleshooting section and add ``--save_lp`` to the ``calliope run`` command to make troubleshooting easier
* #193 (partial) -- Add docstring to demand_share_constraint_rule
* #203 / #193 (partial) -- Add documentation for group cost constraints

# Reviewer checklist

- [ ] Test(s) added to cover contribution
- [x] Documentation updated
- [x] Changelog updated
- [ ] Coverage maintained or improved

### CODE REVIEW

File: calliope/cli.py [hunk 1]
```diff
@@ -172,23 +193,63 @@ def new(path, template, debug):
...
 @click.option('--scenario')
+@click.option('--model_format')
+@click.option('--override_dict')
 @click.option('--save_netcdf')
 @click.option('--save_csv')
 @click.option('--save_plots')
 @click.option('--save_logs')
-@click.option('--model_format')
-@click.option('--override_dict')
+@click.option(
```
Comment (brynpickering) [inline calliope/cli.py]: Any reason to not put in the help strings for the other arguments?

File: calliope/example_models/national_scale/scenarios.yaml [hunk 2]
```diff
@@ -2,8 +2,8 @@
 # Scenarios are optional, named combinations of overrides
```
Comment (brynpickering) [inline calliope/example_models/national_scale/scenarios.yaml]: Do we need these scenarios any more, given that we have our minimal examples?

File: calliope/test/test_constraint_results.py [hunk 3]
```diff
@@ -8,7 +8,7 @@
 
 
 class TestNationalScaleExampleModelSenseChecks:
```
Comment (brynpickering) [inline calliope/test/test_constraint_results.py]: Again, these are essentially duplicating tests now given by the minimal group constraint models, do we need them?

File: calliope/cli.py [hunk 4]
```diff
@@ -172,23 +193,63 @@ def new(path, template, debug):
...
-@cli.command(short_help='Run a model.')
+def _run_setup_model(
+        model_file, scenario, model_format, override_dict):
+    """
+    Build model in CLI commands. Returns ``model``, a ready-to-run
+    calliope.Model instance.
+
+    """
+    # Try to determine model file type if not given explicitly
+    if model_format is None:
```
Comment (brynpickering) [inline calliope/cli.py]: Is this tested?

File: doc/user/troubleshooting.rst [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+* Try setting ``run.solver_options.Presolve: 0``, as large numeric ranges can cause the pre-solver to generate an infeasible or numerically unstable model.
+
+.. seealso::
+
+    The `Gurobi Guidelines for Numerical Issues <https://www.gurobi.com/documentation/current/refman/numerics_gurobi_guidelines.html>`_ give detailed guidance for strategies to address numerically difficult optimisation problems.
+
+Debugging model errors
+----------------------
+
+Calliope provides a method to save its fully built and commented internal representation of a model to a single YAML file with ``Model.save_commented_model_yaml(path)``. Comments in the resulting YAML file indicate where original values were overridden.
```
Comment (brynpickering) [inline doc/user/troubleshooting.rst]: Does this not depend on us sticking with the current method of commented dicts, which depends on sticking with ruamel.yaml?

File: doc/user/troubleshooting.rst [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+To deal with numerically unstable models:
+
+* Try setting ``run.solver_options.Presolve: 0``, as large numeric ranges can cause the pre-solver to generate an infeasible or numerically unstable model.
+
+.. seealso::
+
+    The `Gurobi Guidelines for Numerical Issues <https://www.gurobi.com/documentation/current/refman/numerics_gurobi_guidelines.html>`_ give detailed guidance for strategies to address numerically difficult optimisation problems.
+
+Debugging model errors
```
Comment (brynpickering) [inline doc/user/troubleshooting.rst]: Perhaps also include `model.backend.access_pyomo_model_inputs`, since that will also include NaN values filled with defaults, basically the final parameter sets that pyomo will draw values from (another reason that group constraints should follow a default convention alongside the other constraints?)

File: doc/user/troubleshooting.rst [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+* **Saving an LP file without running the model**: The LP file contains the mathematical model formulation of a fully built Calliope model. It is a standard format that can be passed to various solvers. Examining the LP file manually or using additional tools (see below) can help find issues when a model is infeasible or unbounded. To build a model and save it to LP without actually solving it, use:
+
+  .. code-block:: shell
+
+    calliope run my_model.yaml --save_lp=my_saved_model.lp
+
+Understanding infeasibility and numerical instability
+-----------------------------------------------------
+
+Using the Gurobi solver
```
Comment (brynpickering) [inline doc/user/troubleshooting.rst]: I'll write something about CPLEX too

File: doc/user/building.rst [hunk 8]
```diff
@@ -155,6 +155,50 @@ This will create an ``unmet_demand`` decision variable in the optimisation, whic
...
+        essentials:
+            name: 'Rooftop PV'
+            color: '#B59C2B'
+            parent: supply
+            carrier_out: power
+        constraints:
+            resource: file=pv_resource.csv
+            energy_cap_max: 10000  # kW
+
+All time series data in a model must be indexed by ISO 8601 compatible time stamps (usually in the format ``YYYY-MM-DD hh:mm:ss``, e.g. ``2005-01-01 00:00:00``), i.e., the first column in the CSV file must be time stamps.
```
Comment (brynpickering) [inline doc/user/building.rst]: Not strictly true, a user can also define the timestamp format that their data is stored in...

File: calliope/cli.py (hunk 1 above)
Comment (sjpfenninger) [reply calliope/cli.py]: No, but these were already missing before, and this PR already does a lot of stuff ;)

File: calliope/example_models/national_scale/scenarios.yaml (hunk 2 above)
Comment (sjpfenninger) [reply calliope/example_models/national_scale/scenarios.yaml]: I would leave them until we actually have consistent minimal examples -- right now, the minimal examples are only visible if you actually dig through the source code looking at the tests (and who does _that_?)

File: doc/user/troubleshooting.rst (hunk 6 above)
Comment (sjpfenninger) [reply doc/user/troubleshooting.rst]: can you add that?

File: doc/user/troubleshooting.rst (hunk 5 above)
Comment (sjpfenninger) [reply doc/user/troubleshooting.rst]: This was just moved from where it previously lived in the documentation, nothing new actually implemented

File: calliope/test/test_constraint_results.py (hunk 3 above)
Comment (sjpfenninger) [reply calliope/test/test_constraint_results.py]: These tests actually picked up issues with the group constraints that the group constraint tests did not, so they are not duplicating them (again, the group constraints need better tests before 0.6.4 release)

File: doc/user/building.rst (hunk 8 above)
Comment (sjpfenninger) [reply doc/user/building.rst]: Addressed in ba3f4c5

**Comment by sjpfenninger** [comment]
Updated in 5459ca3 to make default configuration more accessible and visible in the docs. To fully address #161, we would also empty the `required_constraints` lists for all abstract base techs, except leaving `resource` required for `demand` techs. I will hold off on that until we are clear we want to, as it will require updating / fixing a few tests. Does that sound sensible?

File: calliope/test/test_constraint_results.py (hunk 3 above)
Comment (brynpickering) [reply calliope/test/test_constraint_results.py]: Yeah, I'd prefer that the group constraint tests are kept to their respective minimal examples, and that we don't have to pick up strange edge cases using the national-scale / urban-scale example models... If nothing else, it should keep test time down

**Comment by brynpickering** [comment]
I can see some overlap with the new things you've added and existing documentation on improving solution times (https://calliope.readthedocs.io/en/latest/user/running.html#improving-solution-times). I wonder if they should just be cross-referenced, or whether the information from one should be wholly moved to the location of the other?

File: calliope/test/test_constraint_results.py (hunk 3 above)
Comment (sjpfenninger) [reply calliope/test/test_constraint_results.py]: We should consider this when finishing up #193

**Comment by sjpfenninger** [comment]
> I can see some overlap with the new things you've added and existing documentation on improving solution times (https://calliope.readthedocs.io/en/latest/user/running.html#improving-solution-times). I wonder if they should just be cross-referenced, or whether the information from one should be wholly moved to the location of the other?

Good point - done.

**Comment by sjpfenninger** [comment]
@brynpickering I've removed most required constraints -- are we happy with that? This also removes the pre-processing check for "at least one constraint must be defined" (which wasn't actually checking for that, it was only checking for required constraints, not constraints generally). There could be smarter checks to add back in, but there are now so many ways in which a tech can be constrained, that it might be easiest to just leave it up to the user to ensure that what they doing is sensible.

**Comment by brynpickering** [comment]
@sjpfenninger Made some updates to docs and to tests. Couple of things:

1. Not a fan of how inline code snippets look. Any white space is missed in the highlighting... 
image

2. Catching exceptions in CLI seems a bit difficult, I've implemented some tests to catch some excepted ValueErrors, but the current method is to hoover up all errors and then print it with a sys exit code of 1. If you know how to tap into that printout to check for the actual errors being tested, that would be good. At the moment, they could just exiting with code of 1 because the files being used don't exist (and not because they should fail before knowing that).
